Rational PoSt (#332)
filecoin-project/specs: actors.md

	## when a PoSt is submitted (not as each new sector commitment is added).
    provingSet &SectorSet

>     ## Faulty sectors reported since last SubmitPost, up to the current PoSt challenge time.

Reviewer: ```suggestion
    ## Faulty sectors reported since last SubmitPost, up to the current proving period's challenge time.
```

    currentFaultSet BitField

>     ## Faults for the next round of PoSt.

Reviewer: We can be more clear: "Faults submitted after the current proving period's challenge time but before the PoSt for that period is submitted. These become the `currentFaultSet` when a PoSt is submitted."

These fields are still a bit hard to describe in the period between when a SubmitPoSt lands and when the subsequent proving period opens, but I think we have the same idea.

	## when a PoSt is submitted (not as each new sector commitment is added).
    provingSet &SectorSet

>     ## Current self reported faults.

Reviewer: ```suggestion
    ## Faulty sectors reported since last SubmitPost.
```

| `IsLate` | 16 |
| `PaymentVerifyInclusion` | 17 |
| `PaymentVerifySector` | 18 |
| `ReportFaultySectors` | 19 |

> | `RecoverFaultySectors` | 20 |

Reviewer: Remove these two un-specified methods from the table.

| `ChangeWorker` | 13 |
| `IsSlashed` |  14 |
| `IsLate` | 15 |

> | `AddFault` | 16 |

Reviewer 2: you once mentioned that we needed "recoveredFault" or something, do we still need that?

> | `AddFault` | 16 |

Reviewer 3: The document uses `AddFaults` elsewhere - but this says `AddFault`.

		Fatal("not enough collateral")
	}

	// Note: There must exist a unique index in the miner's sector set for each
	// sector ID. The `faults`, `recovered`, and `done` parameters of the

> 	// SubmitPoSt method express indices into this sector set.

Reviewer: Request: could we pick a consistent capitalisation of `SubmitPo[sS]t`?

Author: PoSt is the default, but hard to keep correct everywhere

```sh
type SubmitPost struct {
    proofs PoStProof
    doneSet Bitfield
} representation tuple
```

**Algorithm**

```go
func SubmitPost(proofs PoStProof, doneSet Bitfield) {
	if msg.From != self.Worker {
		Fatal("not authorized to submit post for miner")
	}

	feesRequired := 0
    nextProvingPeriodEnd := self.ProvingPeriodEnd + ProvingPeriodDuration(self.SectorSize)

    // TODO: rework fault handling, for now anything later than 2 proving periods is invalid
    if chain.now() > nextProvingPeriodEnd {
        Fatal("PoSt submited too late")
	} else if chain.Now() > self.ProvingPeriodEnd {
		feesRequired += ComputeLateFee(self.power, chain.Now() - self.provingPeriodEnd)
	}

> 	feesRequired += ComputeTemporarySectorFailureFee(self.sectorSize, recovered)

Reviewer: ```suggestion
	feesRequired += ComputeTemporarySectorFailureFee(self.sectorSize, self.currentFaultSet)
```

```sh
type SubmitPost struct {
    proofs PoStProof
    doneSet Bitfield
} representation tuple
```

**Algorithm**

```go
func SubmitPost(proofs PoStProof, doneSet Bitfield) {

> 	if msg.From != self.Worker {

Reviewer 3: Nit: Indentation.

		TransferFunds(msg.From, msg.Value-feesRequired)
	}

    var seed
    if chain.Now() < self.ProvingPeriodEnd {
      // good case, submitted in time
      seed = GetRandFromBlock(self.ProvingPeriodEnd - POST_CHALLENGE_TIME)
    } else {
      // bad case, submitted late, need to take new proving period end as reference
      seed = GetRandFromBlock(nextPovingPeriodEnd - POST_CHALLENGE_TIME)

>     }

Reviewer: There are some obvious cases where this could fail if `SubmitPoSt` is submitted too early, or races against the proving period end arriving. I guess they're implicitly fatal, but the spec should eventually be specific.

		TransferFunds(msg.From, msg.Value-feesRequired)
	}

    seed := GetRandFromBlock(self.ProvingPeriodStart + ProvingPeriodDuration(self.SectorSize) - POST_CHALLENGE_TIME)
    faultSet := self.faultSet

> 	if !VerifyPoSt(self.SectorSize, self.provingSet, seed, proof, faultSet) {

Reviewer: Something somewhere should document that this will fail if a fault has been declared at any point since `self.ProvingPeriodStart + ProvingPeriodDuration(self.SectorSize) - POST_CHALLENGE_TIME`

In case a miner does experience a fault after the challenge window starts, what mechanism will move their proving period forward to the subsequent window so that they can try again with a new fault set?
This is currently inconsistent with code higher up that prevents any PoSt submission after the generation attack time.

Author: This was actually tricker than I thought, I added more words, as well as more tracking to make this correct. please review again

		TransferFunds(msg.From, msg.Value-feesRequired)
	}

>     seed := GetRandFromBlock(self.ProvingPeriodStart + ProvingPeriodDuration(self.SectorSize) - POST_CHALLENGE_TIME)

Reviewer: ```suggestion
    seed := GetRandFromBlock(self.ProvingPeriodEnd - POST_CHALLENGE_TIME)
```

The miner state doesn't have `ProvingPeriodStart`.

>     seed := GetRandFromBlock(self.ProvingPeriodStart + ProvingPeriodDuration(self.SectorSize) - POST_CHALLENGE_TIME)

Reviewer 4: If this is the `GetRandFromBlock` defined below in the mining.md doc, then this is incorrect. the randomness needs to come from exactly the block at `ProvingPeriodEnd - POST_CHALLENGE_TIME`. Any lookback will give attackers an advantage and degrade the security argument that RationalPoSt makes.

Reviewer 4: BLOCKING ^

Author: @sternhenri and I thought actually that the lookback was good to avoid another issue he was concerned about.

Author: also how does the lookback give attackers an advantage? could you explain?

Author: I have thought about this some more and you are right, we need to take the challenge directly from this block, as otherwise it can be precomputed in advance.

Author: fixed

    faultSet := self.currentFaultSet

	if !VerifyPoSt(self.SectorSize, self.provingSet, seed, proof, faultSet) {
		Fatal("proof invalid")
	}

    // The next fault set becomes the current one
    self.currentFaultSet = self.nextFaultSet
    self.nextFaultSet = EmptySectorSet()

>     // TODO: penalize for faults

Reviewer 4: too dooo

Author: yeah, this will stay for now and a new issue for handling post faults will be created

		TransferFunds(msg.From, msg.Value-feesRequired)
	}

    seed := GetRandFromBlock(self.ProvingPeriodStart + POST_CHALLENGE_TIME)

>     faultSet := self.faultSet

Reviewer 3: Nit: `self.` is a Rust idiom, but you're in a `go` code block.

Author: @whyrusleeping  and I agreed to use `self` to refer to the current actor in the pseudo code. should document this somewhere

Reviewer: `self` is not used consistently below, `miner` appears too.

Author: correct, but I am trying to keep changes minimal

Author: correct, but I am trying to keep changes minimal

Author: cleaned up

> 	if !VerifyPost(self.SectorSize, self.provingSet, seed, proof, faultSet) {

Reviewer 3: Nit: This should be `VerifyPoSt` (note capitalization).

		Fatal("proof invalid")
	}

    // clear fault set
    self.faultSet = EmptySectorSet()

    // TODO: penalize for faults

	// Remove doneSet from the current sectors
	miner.Sectors.Subtract(doneSet)

	// Update miner power to the amount of data actually proved during the last proving period.
	oldPower := miner.Power

> 	miner.Power = (miner.ProvingSet.Size() - faultSet.Count()) * miner.SectorSize

Reviewer 3: You just `// clear fault set` above, so the `- faultSet.Count()` isn't going to do much.

Author: faultSet is a local copy, I clear `self.faultSet`

Reviewer 3: Ah hah. Thanks.

Reviewer: IMO the state updates could be re-arranged into a clearer block at the end of the method:
```
// Update miner power to the amount of data actually proved during the last proving period.
self.Power = (self.ProvingSet.Size() - self.FaultSet.Count()) * self.SectorSize

// Done sectors no longer need to be proven
self.Sectors.Subtract(doneSet)
self.ProvingSet = self.Sectors

// Faults are assumed temporary unless the miner has also declared the sector "done"
self.FaultSet = EmptySectorSet()

... // deal state
```

}
```

#### `AddFaults`

**Parameters**

```sh

> type SubmitPost struct {

Reviewer 4: wrong type name here

}
```

#### `AddFaults`

**Parameters**

```sh
type AddFaults struct {
    faults FaultSet
} representation tuple
```

**Algorithm**

```go

> func AddFaults(faults FaultSet) {

Reviewer 4: I actually think we should simplify this all the way to just:
```
func AddFaults(faults FaultSet) {
  self.currentFaultSet = Merge(self.currentFaultSet, faults)
}
```

If you submit faults before you submit your post, they apply to the post you still havent submitted.

Author: If we simplify this, it means you are not allowed to submit any faults between challenge time and post submission. Which becomes an issue if you miss a post, as you can't post new faults anymore.

Reviewer 5: This would allow a miner to accept storage deals, commit and then discard sectors containing the storage, and then add faults each proving period thereafter. Since the fault set uses RLE, the cost of a the AddFault message is constant over a large number of sectors, and the cost per sector drops without bound. At some point the gas costs for faulting on all deals will drop below the value of the deals.

We're going to need some mechanism to limit the number of times a sector can be faulted or at least a mechanism for timing-out faulted sectors so deal arbitration and payment cancellation can occur.

>     let challengeBlockHeight = self.ProvingPeriodStart + ProvingPeriodDuration(self.SectorSize) - POST_CHALLENGE_TIME

Reviewer: ```suggestion
    let challengeBlockHeight = self.ProvingPeriodEnd - POST_CHALLENGE_TIME
```

>     let challengeBlockHeight = self.ProvingPeriodStart + ProvingPeriodDuration(self.SectorSize) - POST_CHALLENGE_TIME

Reviewer 4: ```suggestion
    challengeBlockHeight := self.ProvingPeriodStart + ProvingPeriodDuration(self.SectorSize) - POST_CHALLENGE_TIME
```
